### PR TITLE
Use recorded changes for outputs and plan root output removals

### PR DESCRIPTION
**Pull request by jbardin**
We record output changes in the plan, but don't currently use them for anything other than display. If we have a wholly known output value stored in the plan, we should prefer that for apply in order to ensure consistency with the planned values. This also avoids cases where evaluation during apply cannot happen correctly, like when all resources are being removed or we are executing a destroy.

Recording all change in the state also lets us remove the need for the special destroy root output transformer. That was always a stop-gap until we could plan the outputs, since it was the last usage of the apply graph `Destroy` flag.

`Before` values are now stored correctly in the plan changes too, and the change `Action` is updated to use `NoOp` when there is no change,  and `Update` when the `Before` value was not null. This PR does not yet go into updating the plan rendering, which likely has some changes that could be made to make use of the new more detailed change entries.

### CODE REVIEW

**Comment by jbardin** [comment]
Looking at this again, I don't think it would be much more work to get the proper `Before` values into the changes here while I'm at it. Converting to Draft while I give it a try.

**Comment by apparentlymart** [comment]
It sounds like you're taking another attempt at what I tried in #21762, which does seem more promising now that we have the refresh and plan walks merged. I'm linking this here just in case the chatter I left for myself in comments there is useful in some way, but the relevant code has changed a lot since I worked on that PR so I'm not sure if it will actually be super relevant now.

**Comment by jbardin** [comment]
It's a similar attempt to that PR, thought I didn't actually "separate" the plan and apply values. I've elected to retain the single apply/destroy nodes for planning, as non-root outputs still behave the same between plan and apply. In the process of adding in the before values for the changes, I took the PR a bit further and entirely removed the need for the special `destroyRootOutputTransformer` and the apply destroy flag.

File: terraform/transform_output.go
Comment (mildwonkey) [inline terraform/transform_output.go:75]: we already know destroy is `false` here from the case immediately above, correct? The switch on `case destroy` a couple of lines down seems redundant.

**Review by mildwonkey — COMMENTED** [review]
Overall this looks good! I left a couple of comments, and I did have one concern. We're getting out of my architectural comfort zone and I'd like to see @apparentlymart review this too if possible.

I'm not quite sure how I feel about known `outputs` showing up as `no-op` during a `create` plan (with no state). It doesn't hurt anything; just not quite what I expected. We might want to warn the governance folks about this change in case they are making assumptions based on the `before` values in outputs (I can't imagine _what_ those might be we need to be cautious when making changes to the json plan output - I dream of integration tests 😍 ).

File: terraform/node_output.go
Comment (jbardin) [reply terraform/node_output.go:374]: Thanks, I missed that one! It's not required, since we're not going to find another root output with the same name, but there is no reason to _not_ break here since we found the value we're looking for!

File: terraform/transform_output.go
Comment (jbardin) [reply terraform/transform_output.go:75]: Ah yes, I definitely didn't reduce this all the way. Let me pull that `destroy` logic out and remove the nested switch entirely.

**Comment by jbardin** [comment]
Excellent catch @mildwonkey! It seems during the refactor to fill out the plan completely, the `create` action for known outputs turned into a `no-op`. While it shouldn't hurt anything like you said, I would also like that to be as precise as possible for consumers of the plan itself.

**Review by mildwonkey — APPROVED** [review]
Nice, thank you @jbardin ! It took 2.5 years but I'm finally feeling like I can _usefully_ review these bigger `terraform` changes 😁  Now I feel confident about approving (a second review is never bad though!)

File: terraform/node_output.go
Comment (pselle) [inline terraform/node_output.go:441]: Since you're doing some value work here, this Equals situation probably need to be marks-aware, that is, operating the equality check on unmarked values...

File: terraform/transform_output.go
Comment (pselle) [inline terraform/transform_output.go:50]: This comment doesn't match the following code on my read (it's related to changes, not a root check)

Comment (jbardin) [reply terraform/transform_output.go:50]: you're right, that got stale. I'll move it down closer to the appropriate context.

File: terraform/node_output.go
Comment (jbardin) [reply terraform/node_output.go]: Since this is intended to be used in the output rendering, I've pulled the sensitivity checks back from that work. We don't have any marks to compare here because the state was loaded without them, and even if there were other marks, we currently don't have anything to render if the values are the same. Instead I added specific checks for sensitivity changing between state and config, and set them in Change.

File: terraform/node_output.go
Comment (apparentlymart) [inline terraform/node_output.go:56]: Doesn't `expander.ExpandModule` already take care of this case, by returning just the root module instance if given the root module? :thinking:

**Review by apparentlymart — APPROVED** [review]
This looks good to me!

It looks like this should now be sufficient to simplify the UI-layer handling of output changes, because the `NodeApplyableOutput.setValue` method now has proper access to the prior state and is handling this decision itself:

https://github.com/hashicorp/terraform/blob/af20a769bebb7b083adf85d16963cdf5437624be/backend/local/backend_plan.go#L327-L338

File: terraform/node_output.go
Comment (jbardin) [reply terraform/node_output.go:56]: Thanks. This little snippet is left over from refactoring, and the switch is no longer even relevant since I ended up removing the use of this node entirely for root modules. I'll remove the switch from here before merging.

**Pull request merged**
